src/epochs.lua

```diff
@@ -446,25 +426,30 @@ end
 --- @param observerAddress string The observer address
 --- @param reportTxId string The report transaction ID
 --- @param failedGatewayAddresses table<GatewayAddress> The failed gateway addresses
+--- @param epochIndex number The epoch index
 --- @param timestamp number The timestamp
 --- @return Observations # The updated observations for the epoch
-function epochs.saveObservations(observerAddress, reportTxId, failedGatewayAddresses, timestamp)
+function epochs.saveObservations(observerAddress, reportTxId, failedGatewayAddresses, epochIndex, timestamp)
```

Comment:
This seems like a good update - i.e. ensure that the reported epochIndex has arrived at an appropriate timestamp. Is it possible we weren't enforcing this adequately before? e.g. allowing a previous epoch's report to be submitted in a later epoch?

Reply:
yeah, that's what caught my eye. I believe `Timestamp` is a protected tag by the MU, so it'll throw it out/error if you try and stub it. but the way we were computing the boundaries for the epoch seemed lacking.